### PR TITLE
Clean variants [or dry run clean variants] and estimate freed space on disk

### LINKED ISSUE #2200: Clean variants #1 -> CLI function
A cli command that allows to:

- clean variants from cases (case-id or status)
- Keep variants with actions (comment, causative, evaluated, pinned)
- archived, solved, older than n years.
- Write a notification on the case page.

### PR DESCRIPTION
**Pull request by northwestwitch**
fix #2200
Allows to remove variants from one case or all cases thus freeing up space in the database.

Options:
image

Default status for the options above: 
- default values for status is: ["solved", "archived", "migrated"]
- older-than : 0 # Remove from any case, regardless of how old the analysis is
- min_rank-threshold: 5
- variants-threshold: doesn't have a default (not provided)

I haven't written tests yet because I could use some feedback over the functionality first.. Like did I miss anything or would it be better to change any default value?

**How to test**:
1. Test locally with the demo instance. Better by loading more than one case, for instance load also the other RD case and the cancer case:
- scout --demo load case scout/demo/643595.config.yaml
- scout --demo load case scout/demo/cancer.load_config.yaml

- Run the command: `scout --loglevel INFO  --demo delete variants` + options
note that the demo data has status=inactive by default, so either you change the status of the case in the web page or you provide the -status inactive param using the delete command above

1. Then I would do also a dry run on scout stage on hasta.

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [x] tests executed by CR

### CODE REVIEW

**Comment by northwestwitch** [comment]
@moonso and @dnil I could use some feedback on this one. It's fine without a proper review for the time being since I haven't even written the tests yet. But if you have comments to improve it it would be great!

**Comment by northwestwitch** [comment]
In the meanwhile I've improved it and now it's ready for review!

**Comment by northwestwitch** [comment]
I'm trying to test this PR on hasta stage but looks like it has installed pymongo==3.6.1 (Scout current master requires pymongo>=3.7) and that generates this error:

image

**Comment by northwestwitch** [comment]
Testing on clinical-db and looks like it's working, but at a certain point I get this error:

```
Remove 31437 / 33855 total variants
#### 94/120 ###### F0010652 (cust002-F0010652)
Remove 24358 / 26199 total variants
#### 96/120 ###### F0006489 (cust002-F0006489)
Remove 23099 / 25067 total variants
#### 97/120 ###### F0009777 (cust002-F0009777)
Remove 28589 / 30884 total variants
#### 98/120 ###### F0008314 (cust002-F0008314)
Remove 23685 / 25645 total variants
#### 99/120 ###### F00042779_2 (cust002-F00042779_2)
Remove 17635 / 20310 total variants
#### 100/120 ###### F0011655 (cust002-F0011655)
Remove 12886 / 13607 total variants
Traceback (most recent call last):
  File "/home/hiseq.clinical/miniconda//envs/prod-stage/bin/scout", line 8, in <module>
    sys.exit(cli())
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/flask/cli.py", line 586, in main
    return super(FlaskGroup, self).main(*args, **kwargs)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/flask/cli.py", line 426, in decorator
    return __ctx.invoke(f, *args, **kwargs)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/scout/commands/delete/delete_command.py", line 54, in variants
    for nr, case in enumerate(cases, 1):
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/cursor.py", line 1207, in next
    if len(self.__data) or self._refresh():
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/cursor.py", line 1144, in _refresh
    self.__send_message(g)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/cursor.py", line 1001, in __send_message
    address=self.__address)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/mongo_client.py", line 1372, in _run_operation_with_response
    exhaust=exhaust)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/mongo_client.py", line 1471, in _retryable_read
    return func(session, server, sock_info, slave_ok)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/mongo_client.py", line 1366, in _cmd
    unpack_res)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/server.py", line 137, in run_operation_with_response
    first, sock_info.max_wire_version)
  File "/home/hiseq.clinical/miniconda/envs/prod-stage/lib/python3.6/site-packages/pymongo/helpers.py", line 164, in _check_command_response
    max_wire_version)
pymongo.errors.CursorNotFound: Cursor not found, cursor id: 162935484061, full error: {'ok': 0.0, 'errmsg': 'Cursor not found, cursor id: 162935484061', 'code': 43}
```

I'll try to fix it, and perhaps speed up the queries a bit!

**Comment by northwestwitch** [comment]
This is tested on clinical-db stage and it's working now. Performed a -dry run with the following options:
Delete variants: 
- older than 4 years
- with rank score < 10
- for cases with status: ["solved", "archived", "migrated"]

It finds 120 cases filtering by the parameters above

Command:
`scout --loglevel WARNING delete variants -older-than 48 -rank-threshold 10 -variants-threshold 20000 --dry-run
`

During the script it applies the other option -> remove variants only for cases with more than 20K variants

The script run in about 45 minutes (It's a lot, I know, but it's a huge number of documents. I've tried to optimize it but without success).

Result:

```
--------------- DRY RUN COMMAND ---------------
#### Case n. 2/120 ###### F0008129 (cust002-F0008129)
Remove 65320 / 66089 total variants
#### Case n. 3/120 ###### F0008070 (cust002-F0008070)
Remove 21358 / 22999 total variants
#### Case n. 4/120 ###### 17033 (cust003-17033)
Remove 34207 / 36479 total variants
#### Case n. 7/120 ###### 216701600274fam (cust004-216701600274fam)
Remove 28805 / 39690 total variants
#### Case n. 8/120 ###### F0007276 (cust002-F0007276)
Remove 30092 / 32418 total variants
#### Case n. 12/120 ###### F0012228 (cust002-F0012228)
Remove 23235 / 24849 total variants
#### Case n. 14/120 ###### F0011383 (cust002-F0011383)
Remove 77261 / 81679 total variants
#### Case n. 20/120 ###### F0005963 (cust002-F0005963)
Remove 22427 / 24141 total variants
#### Case n. 21/120 ###### F0011979 (cust002-F0011979)
Remove 29173 / 31396 total variants
#### Case n. 25/120 ###### 17025 (cust003-17025)
Remove 29649 / 31737 total variants
#### Case n. 27/120 ###### F0011838 (cust002-F0011838)
Remove 22957 / 24570 total variants
#### Case n. 36/120 ###### 516701600590fam (cust004-516701600590fam)
Remove 24087 / 26071 total variants
#### Case n. 37/120 ###### 016701600591fam (cust004-016701600591fam)
Remove 35903 / 47740 total variants
#### Case n. 38/120 ###### 917701600087fam (cust004-917701600087fam)
Remove 23672 / 25594 total variants
#### Case n. 47/120 ###### 417701600088fam (cust004-417701600088fam)
Remove 24684 / 26682 total variants
#### Case n. 49/120 ###### 17017 (cust003-17017)
Remove 34181 / 36010 total variants
#### Case n. 50/120 ###### 17019 (cust003-17019)
Remove 18966 / 20014 total variants
#### Case n. 52/120 ###### F0011960 (cust002-F0011960)
Remove 37120 / 39578 total variants
#### Case n. 53/120 ###### 917701600103fam (cust004-917701600103fam)
Remove 22319 / 24036 total variants
#### Case n. 54/120 ###### 316701600462 (cust004-316701600462)
Remove 20411 / 21966 total variants
#### Case n. 60/120 ###### F0005756 (cust002-F0005756)
Remove 20302 / 21635 total variants
#### Case n. 61/120 ###### F0006358 (cust002-F0006358)
Remove 21717 / 23449 total variants
#### Case n. 62/120 ###### F0009130 (cust002-F0009130)
Remove 32137 / 34275 total variants
#### Case n. 64/120 ###### 517701600077fam (cust004-517701600077fam)
Remove 20896 / 22771 total variants
#### Case n. 65/120 ###### 617701600067fam (cust004-617701600067fam)
Remove 25718 / 27627 total variants
#### Case n. 66/120 ###### 17009 (cust003-17009)
Remove 26303 / 28513 total variants
#### Case n. 67/120 ###### 17014 (cust003-17014)
Remove 38540 / 41172 total variants
#### Case n. 69/120 ###### 117701600076fam (cust004-117701600076fam)
Remove 30192 / 32544 total variants
#### Case n. 74/120 ###### 17007 (cust003-17007)
Remove 31789 / 34293 total variants
#### Case n. 77/120 ###### 17003 (cust003-17003)
Remove 24841 / 26635 total variants
#### Case n. 78/120 ###### 17005 (cust003-17005)
Remove 24897 / 26206 total variants
#### Case n. 79/120 ###### 517701600028fam (cust004-517701600028fam)
Remove 25245 / 27189 total variants
#### Case n. 80/120 ###### F0011579 (cust002-F0011579)
Remove 20122 / 21583 total variants
#### Case n. 81/120 ###### 117701600068fam (cust004-117701600068fam)
Remove 22751 / 24497 total variants
#### Case n. 82/120 ###### F0011768 (cust002-F0011768)
Remove 19187 / 20205 total variants
#### Case n. 83/120 ###### F0007735 (cust002-F0007735)
Remove 28721 / 30893 total variants
#### Case n. 86/120 ###### 16172 (cust003-16172)
Remove 27967 / 30195 total variants
#### Case n. 88/120 ###### F0007736 (cust002-F0007736)
Remove 223883 / 225492 total variants
#### Case n. 89/120 ###### F0011639 (cust002-F0011639)
Remove 137511 / 138774 total variants
#### Case n. 90/120 ###### F0011436 (cust002-F0011436)
Remove 24571 / 26534 total variants
#### Case n. 93/120 ###### F0007485 (cust002-F0007485)
Remove 31437 / 33855 total variants
#### Case n. 94/120 ###### F0010652 (cust002-F0010652)
Remove 24358 / 26199 total variants
#### Case n. 96/120 ###### F0006489 (cust002-F0006489)
Remove 23099 / 25067 total variants
#### Case n. 97/120 ###### F0009777 (cust002-F0009777)
Remove 28589 / 30884 total variants
#### Case n. 98/120 ###### F0008314 (cust002-F0008314)
Remove 23685 / 25645 total variants
#### Case n. 99/120 ###### F00042779_2 (cust002-F00042779_2)
Remove 17635 / 20310 total variants
#### Case n. 102/120 ###### F0008158 (cust002-F0008158)
Remove 19667 / 22651 total variants
#### Case n. 104/120 ###### F0010682 (cust002-F0010682)
Remove 18604 / 21635 total variants
#### Case n. 106/120 ###### F0010096 (cust002-F0010096)
Remove 24669 / 26649 total variants
#### Case n. 107/120 ###### 16701600609fam (cust004-16701600609fam)
Remove 23204 / 25060 total variants
#### Case n. 109/120 ###### F0006366 (cust002-F0006366)
Remove 26357 / 28435 total variants
#### Case n. 113/120 ###### F0010718 (cust002-F0010718)
Remove 31542 / 33916 total variants
#### Case n. 117/120 ###### 16153 (cust003-16153)
Remove 27930 / 29224 total variants
Total estimated deleted variants: 1773893
Estimated space freed (GB): 24.780998937785625
```

**Comment by northwestwitch** [comment]
Perhaps it would be nice to create an event associated to the variant cleaning, whenever variants get removed from a case, so the users will know!

**Comment by northwestwitch** [comment]
I've added the code to save an event when a variants get removed:

image

**Comment by northwestwitch** [comment]
This is a bit urgent, since we need to merge it in order to run a simulation to see how many variants would be removed and from which samples if we want to clean the database before we move it to the new server. Do you have time to review @moonso ?

File: scout/adapter/mongo/base.py [hunk 1]
```diff
@@ -120,5 +121,17 @@ def collections(self):
         """
         return self.db.list_collection_names()
 
+    def collection_stats(self, coll_name):
```
Comment (moonso) [inline scout/adapter/mongo/base.py:124]: Maybe I'm a bit biased now but it is really nice to use the `typing` types and utilise pythons type hints functionality. It becomes more easy to program since you get a lot for free with the editor (at least with PyCharm). One can also use [mypy](https://mypy.readthedocs.io/en/stable/) to check that input and output is correctly defined. So in this case: 
```suggestion
    def collection_stats(self, coll_name: str) -> dict:
```

File: scout/adapter/mongo/base.py [hunk 2]
```diff
@@ -120,5 +121,17 @@ def collections(self):
...
+        """Returns stats from a single collection
+
+        Args:
+            coll_name: name of collection to retrieve stats for
+
+        Returns:
+            stats(dict): dictionary with collection stats
+        """
+        db = current_app.config.get("MONGO_DATABASE")
+        return db.command("collstats", coll_name)
```
Comment (moonso) [inline scout/adapter/mongo/base.py:134]: Does this really return a dictionary or is it a pymongo object?

File: scout/adapter/mongo/case_events.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ def assign(self, institute, case, user, link):
         This function will create an Event to log that a person has been assigned
         to a case. Also the user will be added to case "assignees".
 
-        Arguments:
+        Args:
```
Comment (moonso) [inline scout/adapter/mongo/case_events.py:21]: Again I would recommend moving towards type hints

File: scout/adapter/mongo/case_events.py [hunk 4]
```diff
@@ -225,10 +225,30 @@ def archive_case(self, institute, case, user, link):
         LOG.debug("Case updated")
         return updated_case
 
+    def remove_variants_event(self, institute, case, user, link, content):
```
Comment (moonso) [inline scout/adapter/mongo/case_events.py:228]: 👍

File: scout/adapter/mongo/query.py [hunk 5]
```diff
@@ -13,6 +14,54 @@
 
 
 class QueryHandler(object):
+    def build_case_query(self, case_id=None, status=None, older_than=None):
```
Comment (moonso) [inline scout/adapter/mongo/query.py:17]:
```suggestion
    def build_case_query(self, case_id=None, status=None, older_than=None) -> dict:
```

File: scout/adapter/mongo/query.py [hunk 6]
```diff
@@ -13,6 +14,54 @@
...
+        """
+        variants_query = {}
+        case_subquery = {"case_id": case_id}
+
+        # Create query to delete all variants that shouldn't be kept or with rank higher than min_rank_threshold
+        if variants_to_keep or min_rank_threshold:
+            variants_query["$and"] = [case_subquery]
+            if variants_to_keep:
+                variants_query["$and"].append({"_id": {"$nin": variants_to_keep}})
+            if min_rank_threshold:
```
Comment (moonso) [inline scout/adapter/mongo/query.py:58]: Do you know how this will work for cases without rank scores? I know there are some cancer cases with millions of variants without rank score. Could be worth testing...

File: scout/commands/delete/delete_command.py [hunk 7]
```diff
@@ -8,6 +8,106 @@
...
+
+    user_obj = store.user(user)
+    if user_obj is None:
+        click.echo(f"Could not find a user with email '{user}' in database")
+        return
+
+    total_deleted = 0
+    items_name = "deleted variants"
+    if dry_run:
+        click.echo("--------------- DRY RUN COMMAND ---------------")
```
Comment (moonso) [inline scout/commands/delete/delete_command.py:49]:
```suggestion
        LOG.info("--------------- DRY RUN COMMAND ---------------")
```

File: scout/commands/delete/delete_command.py [hunk 8]
```diff
@@ -8,6 +8,106 @@
...
+    # Estimate the average size of a variant document in database
+    avg_var_size = store.collection_stats("variant").get("avgObjSize", 0)  # in bytes
+
+    # Get all cases where case_query applies
+    n_cases = store.case_collection.count_documents(case_query)
+    cases = store.cases(query=case_query)
+    filters = (
+        f"Rank-score threshold:{rank_threshold}, case n. variants threshold:{variants_threshold}."
+    )
+    for nr, case in enumerate(cases, 1):
```
Comment (moonso) [inline scout/commands/delete/delete_command.py:65]: Suggest to put the logic here in a function and let the cli only handle input and output. Makes easier to refactor into smaller functions

File: scout/commands/delete/delete_command.py (hunk 7 above)
Comment (moonso) [reply scout/commands/delete/delete_command.py:49]: Only a suggestion, do as you prefer. It makes sense if you refactor this to own functions to use logging

File: scout/commands/delete/delete_command.py [hunk 9]
```diff
@@ -8,6 +8,106 @@
...
+            url = url_for(
+                "cases.case", institute_id=institute_obj["_id"], case_name=case["display_name"]
+            )
+            store.remove_variants_event(
+                institute=institute_obj, case=case, user=user_obj, link=url, content=filters
+            )
+
+    click.echo(f"Total {items_name}: {total_deleted}")
+    click.echo(
+        f"Estimated space freed (GB): {round((total_deleted * avg_var_size) / 1073741824, 4)}"
```
Comment (moonso) [inline scout/commands/delete/delete_command.py:107]: 😄

Comment (moonso) [reply scout/commands/delete/delete_command.py:107]: Perhaps this number could be a variable given in the CLI instead of hard coded. If we do this multiple times

**Review by moonso — APPROVED** [review]
Great!! I had some comments but nothing blocking

File: scout/commands/delete/delete_command.py (hunk 9 above)
Comment (northwestwitch) [reply scout/commands/delete/delete_command.py:107]: Not on the CLI but perhaps as a constant. It's 1024*1024*1024, to convert the bytes in GB

**Comment by northwestwitch** [comment]
Thanks for the review @moonso, I'll fix accordingly.

@henrikstranneheim and I had a chat about how to use the script, once it's merged.

We proposed to filter by:
- Cases older than 1 year
- Cases with >= 50K variants
- Cases with status solved, archived, migrated
- Variant score <10

I'd first run a simulation over the weekend.

Any suggestion? @dnil?

**Comment by dnil** [comment]
Did you check the 50k cutoff yet, or is that a hunch still? My vague recollection is that that will catch a few "ordinary" patients as well. I would have guessed it tapered of at 75-80k somewhere, at least used to. The 100k sounds like a great start: if we get 1/3 of the variants removed by pruning like 3% of the cases, that is very good. But we will loose analysis quality if we dont retain a decent coverage of the score 5-10 variants. Especially a few years back, we had real issues with the scoring of frameshifts and founders, and we still have trouble with splice variants.

**Comment by northwestwitch** [comment]
> f we get 1/3 of the variants removed by pruning like 3% of the cases

It's not exactly 1/3 of the cases because the cases from the email of this morning contained:
- Perhaps cases from the last year (that would be filtered out from this script, unless we decide to apply to all cases regardless of the analysis date)
- Not all variants are going to disappear from those 100K cases, but most of them for sure.

But sure, we could go for cases with 100K or more

File: scout/commands/delete/delete_command.py (hunk 7 above)
Comment (northwestwitch) [reply scout/commands/delete/delete_command.py:49]: I am a bit hesitant here because I'm not sure how this is going to be written in the log. Specifically:
- This should be runned either on hasta or clinical-db, cli. Does the clinical-db write to a log file? I'm not sure.
- Since the command might take quite some time, what is some other command writes stuff in between on the log file? It would be a mess..

File: scout/commands/delete/delete_command.py (hunk 8 above)
Comment (northwestwitch) [reply scout/commands/delete/delete_command.py:65]: Yes you'd have smaller functions but on the other hand it would become more hard to follow the programming logic, and to understand what the function does you'd have to open another file. When I wrote the delete_variants function it was obscenely complicated and the cyclomatic complexity was very high. I refactored and moved stuff in 3 new functions:

- collection_stats
- build_case_query
- remove_variants_event

Plus it's already calling `evaluated_variants`. 
Now the complexity is fine (codefactor agrees) and it's very easy to understand what is happening. I wouldn't fragment it more than this

File: scout/commands/delete/delete_command.py (hunk 9 above)
Comment (northwestwitch) [reply scout/commands/delete/delete_command.py:107]: Or perhaps I didn't understand your question?

File: scout/adapter/mongo/base.py (hunk 2 above)
Comment (northwestwitch) [reply scout/adapter/mongo/base.py:134]: Dictionary. Using a  type(db.command("collstats", coll_name)) debug message:

image

File: scout/adapter/mongo/base.py (hunk 1 above)
Comment (northwestwitch) [reply scout/adapter/mongo/base.py:124]: Good suggestion, I'll add the typing types!

File: scout/adapter/mongo/case_events.py (hunk 3 above)
Comment (northwestwitch) [reply scout/adapter/mongo/case_events.py:21]: Done, thanks!

File: scout/adapter/mongo/query.py (hunk 6 above)
Comment (northwestwitch) [reply scout/adapter/mongo/query.py:58]: It's a good question! Tested on the demo cancer case with this command:
`scout --loglevel INFO --demo delete variants -u clark.kent@mail.com -c cancer_test -status inactive`

And got some variants removed. But test cancer case does have rank scores: 😆 

image

If you take for instance this case with no rank scores: https://scout.scilifelab.se/cust087/normalcondor, checking any variant in the database it has score=0
image
Then no variant will be removed, unless you specify rank_score=0 in the command params.

But I guess it's good be should be so. We should perhaps establish another system to remove variants from these cases and the'll have to be removed in another way

File: scout/commands/delete/delete_command.py (hunk 9 above)
Comment (northwestwitch) [reply scout/commands/delete/delete_command.py]: Fixed by creating this constant:
`BYTES_IN_ONE_GIGABYTES = 1073741824  # (1024*1024*1024)`

File: scout/commands/delete/delete_command.py (hunk 7 above)
Comment (northwestwitch) [reply scout/commands/delete/delete_command.py:51]: I'll keep it like this for now. We can change it later

**Comment by northwestwitch** [comment]
If nobody has strong objections I'd merge as soon as it passes the tests..
